dagql/dagui/opts.go

```diff
@@ -51,33 +56,46 @@ const (
 	ShowDigestsVerbosity      = 4
 )
 
-func (opts FrontendOpts) ShouldShow(tree *TraceTree) bool {
+func (opts FrontendOpts) ShouldShow(span *Span) bool {
 	if opts.Debug {
 		// debug reveals all
 		return true
 	}
-	span := tree.Span
-	if span.IsInternal() && opts.Verbosity < ShowInternalVerbosity {
-		// internal steps are hidden by default
-		return false
+	if opts.FocusedSpan == span.ID {
+		// prevent focused span from disappearing
+		return true
 	}
-	if tree.Parent != nil && (span.Encapsulated || tree.Parent.Span.Encapsulate) && tree.Parent.Span.Err() == nil && opts.Verbosity < ShowEncapsulatedVerbosity {
-		// encapsulated steps are hidden (even on error) unless their parent errors
+	if span.Ignore {
+		// absolutely 100% boring spans, like 'id' and 'sync'
+		//
+		// this is ahead of failed check because 'sync' is often failed and is
+		// _still_ not interesting
 		return false
 	}
-	if span.Err() != nil {
-		// show errors
+	if span.IsFailedOrCausedFailure() {
+		// prioritize showing failed things, even if they're internal
 		return true
 	}
-	if tree.IsRunningOrChildRunning {
-		// show running steps
+	if span.Hidden(opts) {
+		return false
+	}
+	if span.IsPending() {
+		// reveal pending spans so the user can see what's queued to run
 		return true
 	}
-	if tree.Parent != nil && (opts.TooFastThreshold > 0 && span.ActiveDuration(time.Now()) < opts.TooFastThreshold && opts.Verbosity < ShowSpammyVerbosity) {
-		// ignore fast steps; signal:noise is too poor
-		return false
+	if span.IsRunningOrLinksRunning() {
+		return true
 	}
-	if opts.GCThreshold > 0 && time.Since(span.EndTime()) > opts.GCThreshold && opts.Verbosity < ShowCompletedVerbosity {
+	// TODO: avoid breaking chains
+	// if opts.TooFastThreshold > 0 &&
+	// 	span.ActiveDuration(time.Now()) < opts.TooFastThreshold &&
+	// 	opts.Verbosity < ShowSpammyVerbosity {
+	// 	// ignore fast steps; signal:noise is too poor
+	// 	return false
+	// }
```

Comment:
Yeah - now that we have a "pending" state we actually _want_ to show most spammy spans, otherwise you wouldn't see the pending state.

I'm going to experiment more with this - I think we probably want to hide entire pipeline chains that took less than some threshold, instead of individual spans.